### PR TITLE
feat(vmbda): apply new controller design

### PR DESCRIPTION
**Pull request by Isteb4k**
## Description

- New vmbda controller design
- Conditions: BlockDeviceReady, VirtualMachineReady, Attached

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/vmbda/vmbda_webhook.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	return nil, nil
+}
+
+func (v *Validator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldCVI, ok := oldObj.(*virtv2.VirtualMachineBlockDeviceAttachment)
+	if !ok {
+		return nil, fmt.Errorf("expected an old VirtualMachineBlockDeviceAttachment but got a %T", newObj)
+	}
+
+	newCVI, ok := newObj.(*virtv2.VirtualMachineBlockDeviceAttachment)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmbda/vmbda_webhook.go:52]: CVI?

File: images/virtualization-artifact/pkg/controller/vm/internal/block_device.go [hunk 2]
```diff
@@ -108,17 +108,51 @@ func (h *BlockDeviceHandler) Handle(ctx context.Context, s state.VirtualMachineS
 	}
 
 	if kvvmi != nil {
+		// Set BlockDeviceRef in the status if the disk exists in KVVMI.
 		for _, ref := range changed.Spec.BlockDeviceRefs {
-			if h.findAttachedBlockDevice(changed, ref) == nil {
-				if abd := h.createAttachedBlockDevice(ref, bdState, kvvmi); abd != nil {
+			if !h.hasBlockDevice(changed.Status.BlockDeviceRefs, ref.Kind, ref.Name) {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/block_device.go:113]: fffk. I also have PR with this

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: crds/virtualmachineblockdeviceattachment.yaml [hunk 3]
```diff
@@ -72,21 +102,21 @@ spec:
...
-                  description: |
-                    A brief description of the cause of the error.
-                failureMessage:
-                  type: string
+                    - "Terminating"
+                observedGeneration:
+                  type: integer
                   description: |
-                    Detailed description of the error.
+                    Represents the .metadata.generation that the status was set based upon.
```
Comment (diafour) [inline crds/virtualmachineblockdeviceattachment.yaml:119]:
```suggestion
                    Resource generation observed during last reconcile.
```

File: images/virtualization-artifact/pkg/controller/service/resource.go [hunk 4]
```diff
@@ -157,3 +154,23 @@ func GetPatchFinalizers(finalizers []string) (client.Patch, error) {
...
+
+func MergeResults(results ...reconcile.Result) reconcile.Result {
+	var result reconcile.Result
+	for _, r := range results {
+		if r.IsZero() {
+			continue
+		}
+		if r.Requeue {
+			return r
+		}
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/service/resource.go:166]: It returns the first requeue without finding requeue with the least RequeueAfter as stated in the following code.

May be you meant requeue without RequeueAfter?

```suggestion
		if r.Requeue && r.RequeueAfter == 0 {
			return r
		}
```

File: images/virtualization-artifact/pkg/controller/vmbda/internal/watcher/vmbda_watcher.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func NewVirtualMachineBlockDeviceAttachmentWatcher(logger *slog.Logger, client client.Client) *VirtualMachineBlockDeviceAttachmentWatcher {
+	return &VirtualMachineBlockDeviceAttachmentWatcher{
+		client: client,
+		logger: logger.With("watcher", "vmbda"),
+	}
+}
+
+func (w VirtualMachineBlockDeviceAttachmentWatcher) Run(mgr manager.Manager, ctr controller.Controller) error {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmbda/internal/watcher/vmbda_watcher.go:45]: Run is an ambiguous name: it overlaps with Run from SharedInformer. Why not Watch?

File: images/virtualization-artifact/pkg/controller/service/resource.go (hunk 4 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/service/resource.go:166]: we don't need to search for requeuAfter with a lower value if the Requeue value is defined.

Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/service/resource.go:166]: this is the correct code

File: images/virtualization-artifact/pkg/controller/vmbda/internal/watcher/vmbda_watcher.go (hunk 5 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vmbda/internal/watcher/vmbda_watcher.go:45]: mb Register?

File: images/virtualization-artifact/pkg/controller/service/resource.go (hunk 4 above)
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/service/resource.go:166]: If RequeueAfter is set to one month here: then the reconciliation will occur only after one month.
`reconcile.Result{Requeue: true, RequeueAfter: 999999999999999 * time.Hour}`
The case with case result.RequeueAfter > 0 is evaluated before the case case result.Requeue.

File: images/virtualization-artifact/pkg/controller/vmbda/internal/block_device_ready.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vmbdacondition.BlockDeviceNotReady
+			condition.Message = fmt.Sprintf("VirtualDisk %s is not Ready: waiting for the VirtualDisk to be Ready.", vdKey.String())
+			return reconcile.Result{}, nil
+		}
+
+		if vd.Status.Target.PersistentVolumeClaim == "" {
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vmbdacondition.BlockDeviceNotReady
+			condition.Message = "Waiting for the underlying PersistentVolumeClaim name will be known."
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmbda/internal/block_device_ready.go:100]:
```suggestion
			condition.Message = "Waiting until VirtualDisk has associated PersistentVolumeClaim name."
```

**Review by yaroslavborbat — APPROVED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**